### ISSUE TITLE
filecopy_example.cpp is buggy

### ISSUE THREAD

**Issue by bravnsgaard** [BoostGSoC13/boost.afio#71]
The *less toy* example has a bug in the parsing of the input files. It loops, incrementing `argc` instead of `n`, as it should.

Even after fixing that I cannot get the program to work, and I am unsure exactly what is wrong. The output file for the concatenated inputs is created and `truncate`d to the correct size, but no data ever gets written to this file. If I use large files I have seen `std::invalid_argument` exceptions and even segfaults. I *think* the problem may be related to the buffers; I have tried using a simple `std::vector<char>` for the `file_buffer_type`, but this makes no difference. 

UPDATE: So, I've been hacking a bit at this, and it turns out the buffer is not the problem. The problem seems to be here:
```c++
// Schedule a filling of buffer from offset o after last has completed
auto readchunk=dispatcher->read(make_async_data_op_req(last, buffer->data(),
    thischunk, o));
// Schedule a writing of buffer to offset offset+o after readchunk is ready
auto writechunk=dispatcher->write(make_async_data_op_req(readchunk,
    buffer->data(), thischunk, offset+o));
```
The precondition for the `write` is the previous `readchunk` future. This holds an io_op to the input file, so we are trying to write to the input file again. This file is opened for reading only, so the operation fails.

**Comment by bravnsgaard** [comment]
I think there is something fundamentally wrong here, most likely resource related. PR #74 fixes the issues mentioned above, but there is still something wrong here.

I have tried using the example to do a simple copy, by "concatenating" just one file into a destination and then doing checksum comparisons between the two resulting files. I get destination files that are zero-padded at the end (some chunks are simply not copied), segmentation faults, and even "FATAL EXCEPTION: Failed to write all buffers", which indicates that we have tried to write past the end of the (correctly sized) destination file. All of this completely random on several consecutive tries.

**Comment by ned14** [comment]
I would be highly surprised if AFIO is broken here. There is a unit test which generates a random sequence of reads and writes, checksumming each time in memory which it repeats on disc, and it checks the sequence produces identical outcomes. That test is extremely sensitive to even a bit being out of place.

Can you send me your failing program please along with what console output it generates? I'm currently doing 14 hour days preparing for C++ Now next week, but I may just be able to see the problem from inspection.

**Comment by bravnsgaard** [comment]
I didn't mean that AFIO is broken, just the example. Sorry if I was unclear. My problem probably is that I am too unversed in AFIO yet. I'll take a look at that unit test and see how it compares.

**Comment by ned14** [comment]
Sorry if I was harsh. A common mistake for both ASIO and AFIO users is using a region of memory in more than one ongoing operation. One needs to never touch a buffer, or any byte in that buffer, after it is handed off to ASIO or AFIO until the operation using that buffer completes.

**Comment by bravnsgaard** [comment]
Right, and that's probably where the issue is here, since all the queueing up happens in a seperate function that allocates the buffers.

You  weren't being harsh, btw. If you're chucking in the hours you say for CppNow, I'd say you have your work cut out for you. Good luck at the conference; I'll have to wait for the videos. :-)

**Comment by ned14** [comment]
You should see my major announcement on boost-dev for the conference shortly which is a handbook of best practices for writing C++ 11/14 (Boost) libraries based on a review of ten upcoming C++ 11/14 libraries (of which four are covered in the C++ Now presentation), and the handbook hyperlinks into the relevant bits of source code. It is the writing of that handbook which currently draws down four hours per day after work and family stuff - indeed, my kid just puked twice over everything which I just cleaned up, so I get to write tonight in quite fetid air :( Still, I need it mostly finished by Friday!

**Comment by ariccio** [comment]
This is a bit off topic: do you run the Windows tests under Application Verifier? Aside from the fact that some of the unit tests timeout/run out of memory, it might save you a few hours if there are any remaining subtle bugs.

I imagine that it'd help track down weird/obscure issues

**Comment by ned14** [comment]
I haven't run the unit tests under Application Verifier since the v1.1 engine. Haven't seen the need - everything passes beautifully almost all of the time, probably because I develop features on Windows first and then POSIX because Windows has true async. Regarding automating its use in the CI, the Windows VM is unfortunately fairly RAM limited. The only known issue on Windows is when running 32 bit processes on a 64 bit kernel, directory enumeration has a 50/50 chance of segfaulting. This is due to a Microsoft bug in Wow6432, I may one day have to disable async enumeration if 32 bit on 64 bit.

**Comment by ariccio** [comment]
> This is due to a Microsoft bug in Wow6432

Now ***that's*** how you know that you've written very good unit/stress tests.

**Comment by ned14** [comment]
Well ... maybe ... it did take me a whole release cycle to figure out the bug. Basically depending on the kernel APC path it parses the wrong 32 bit parameter into its 64 bit location due to a buggy alignment calculation. That parameter is used by ASIO to locate the completion handler, and hence the value can flip flop between any of null, the right value, or pointer to garbage, depending. If it's null ASIO never calls your completion handler, and if it's pointing to garbage you are guaranteed a segfault. I decided on the latter being better than the former.

I tried reporting it to Microsoft, luckily I know people there who got me passed the front line support to the people who actually can fix things in the kernel team. And their response was "that's hugely interesting, but no Microsoft code does async directory enumeration, not anywhere. So it's hard to raise an internal case to fix something nobody uses and aren't supposed to be using until somebody is actually using it, especially as it's been fixed on 64 bit kernels". Oh well.
